### PR TITLE
Changing EntrySet to be based on entries array based on longs instead of ints

### PR DESCRIPTION
**Pull request by sanastas**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: core/src/main/java/com/yahoo/oak/EntrySet.java [hunk 1]
```diff
@@ -190,78 +190,61 @@ int getLastEntryIndex() {
...
             return 0;
         }
         return ((entryIdx - 1) * FIELDS) + HEAD_NEXT_INDEX_SIZE;
     }
 
     /**
+<<<<<<< HEAD
+     * getEntryArrayFieldLong atomically reads long field of the entries array.
+     * Could be used with any OFFSET.value
+=======
```
Comment (liran-funaro) [inline core/src/main/java/com/yahoo/oak/EntrySet.java:210]: Merge issue

File: core/src/main/java/com/yahoo/oak/EntrySet.java [hunk 2]
```diff
@@ -190,78 +190,61 @@ int getLastEntryIndex() {
...
      */
-    private int getEntryArrayFieldInt(int intFieldIdx, OFFSET offset) {
+    private long getEntryArrayFieldInt(int intFieldIdx, OFFSET offset) {
         return entries[intFieldIdx + offset.value]; // used for NEXT
     }
 
     /**
      * getEntryArrayFieldLong atomically reads two integers field of the entries array.
      * Should be used with OFFSET.VALUE_REFERENCE and OFFSET.KEY_REFERENCE
+>>>>>>> 7ea91e5b1433fe70d89c56c5d0ca6614b323f6ab
```
Comment (liran-funaro) [inline core/src/main/java/com/yahoo/oak/EntrySet.java:223]: Merge issue

File: core/src/main/java/com/yahoo/oak/EntrySet.java [hunk 3]
```diff
@@ -190,78 +190,61 @@ int getLastEntryIndex() {
...
      */
     private long getEntryArrayFieldLong(int intStartFieldIdx, OFFSET offset) {
-        assert offset == OFFSET.VALUE_REFERENCE || offset == OFFSET.KEY_REFERENCE;
         long arrayOffset =
-                Unsafe.ARRAY_INT_BASE_OFFSET + (intStartFieldIdx + offset.value) * Unsafe.ARRAY_INT_INDEX_SCALE;
+                Unsafe.ARRAY_LONG_BASE_OFFSET + (intStartFieldIdx + offset.value) * Unsafe.ARRAY_LONG_INDEX_SCALE;
         assert arrayOffset % 8 == 0;
         return UNSAFE.getLong(entries, arrayOffset);
```
Comment (liran-funaro) [inline core/src/main/java/com/yahoo/oak/EntrySet.java:229]: Why not reading directly from the array? Same comment for `setEntryFieldLong()`.
```java
return entries[intStartFieldIdx + offset.value];
```

File: core/src/main/java/com/yahoo/oak/EntrySet.java [hunk 4]
```diff
@@ -190,78 +190,61 @@ int getLastEntryIndex() {
...
+    private long getEntryArrayFieldInt(int intFieldIdx, OFFSET offset) {
         return entries[intFieldIdx + offset.value]; // used for NEXT
     }
 
     /**
      * getEntryArrayFieldLong atomically reads two integers field of the entries array.
      * Should be used with OFFSET.VALUE_REFERENCE and OFFSET.KEY_REFERENCE
+>>>>>>> 7ea91e5b1433fe70d89c56c5d0ca6614b323f6ab
      */
     private long getEntryArrayFieldLong(int intStartFieldIdx, OFFSET offset) {
```
Comment (liran-funaro) [inline core/src/main/java/com/yahoo/oak/EntrySet.java:225]:
```suggestion
    private long getEntryArrayFieldLong(int longStartFieldIdx, OFFSET offset) {
```

File: core/src/main/java/com/yahoo/oak/EntrySet.java [hunk 5]
```diff
@@ -190,78 +190,61 @@ int getLastEntryIndex() {
...
-    /**
-     * setEntryFieldInt sets the integer field of specified offset to 'value'
-     * for given integer index in the entry array
-     */
-    private void setEntryFieldInt(int intFieldIdx, OFFSET offset, int value) {
-        assert intFieldIdx + offset.value >= 0;
-        entries[intFieldIdx + offset.value] = value; // used for NEXT
-    }
-
     private void setEntryFieldLong(int item, OFFSET offset, long value) {
```
Comment (liran-funaro) [inline core/src/main/java/com/yahoo/oak/EntrySet.java:232]:
```suggestion
    private void setEntryFieldLong(int longStartFieldIdx, OFFSET offset, long value) {
```

File: core/src/main/java/com/yahoo/oak/EntrySet.java [hunk 6]
```diff
@@ -190,78 +190,61 @@ int getLastEntryIndex() {
...
+=======
      * getEntryArrayFieldInt gets the integer field of entry at specified offset for given
      * start of the entry index in the entry array.
      * The field is read atomically as it is a machine word, however the concurrency of the
      * mostly updated value is not ensured as no memory fence is issued.
      */
-    private int getEntryArrayFieldInt(int intFieldIdx, OFFSET offset) {
+    private long getEntryArrayFieldInt(int intFieldIdx, OFFSET offset) {
         return entries[intFieldIdx + offset.value]; // used for NEXT
     }
```
Comment (liran-funaro) [inline core/src/main/java/com/yahoo/oak/EntrySet.java:218]: It is currently not in use. I suggest using it when an int field is required.
```suggestion
    private long getEntryArrayFieldInt(int longStartFieldIdx, OFFSET offset) {
        return entries[intFieldIdx + offset.value] & ((1L << 32) - 1); // used for NEXT
    }
```

File: core/src/main/java/com/yahoo/oak/EntrySet.java [hunk 7]
```diff
@@ -309,7 +292,7 @@ int getNextEntryIndex(int ei) {
         if (ei == INVALID_ENTRY_INDEX) {
             return INVALID_ENTRY_INDEX;
         }
-        return getEntryArrayFieldInt(entryIdx2intIdx(ei), OFFSET.NEXT);
+        return (int) getEntryArrayFieldLong(entryIdx2LongIdx(ei), OFFSET.NEXT);
```
Comment (liran-funaro) [inline core/src/main/java/com/yahoo/oak/EntrySet.java:295]: I suggest using here (and below) `getEntryArrayFieldInt()`.
At least use a mask before converting to int:
```suggestion
        return (int) getEntryArrayFieldLong(entryIdx2LongIdx(ei), OFFSET.NEXT) & ((1L << 32) - 1);
```

File: core/src/main/java/com/yahoo/oak/EntrySet.java [hunk 8]
```diff
@@ -593,7 +576,7 @@ boolean deleteValueFinish(ThreadContext ctx) {
 
         assert ctx.valueState == ValueState.DELETED_NOT_FINALIZED;
 
-        int indIdx = entryIdx2intIdx(ctx.entryIndex);
+        int indIdx = entryIdx2LongIdx(ctx.entryIndex);
```
Comment (liran-funaro) [inline core/src/main/java/com/yahoo/oak/EntrySet.java:579]:
```suggestion
        int longIdx = entryIdx2LongIdx(ctx.entryIndex);
```

**Review by liran-funaro — APPROVED** [review]
Looks good to me.

File: core/src/main/java/com/yahoo/oak/EntrySet.java (hunk 3 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/EntrySet.java:229]: good point indeed :)

File: core/src/main/java/com/yahoo/oak/EntrySet.java (hunk 6 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/EntrySet.java:218]: I do not think it worth to keep a function that is not in use. getEntryArrayFieldInt removed for now

File: core/src/main/java/com/yahoo/oak/EntrySet.java (hunk 7 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/EntrySet.java:295]: Int access/assignment is currently done via long to int casting. As other part of long is not used. This is good enough.

**Pull request merged**